### PR TITLE
KAFKA-4486: Don't commit offsets on exception

### PR DESCRIPTION
**Pull request by enothereska**

### CODE REVIEW

**Comment by enothereska** [comment]
Test passed, weird Jenkins error happening more frequently: 
BUILD SUCCESSFUL

Total time: 1 hrs 19 mins 9.575 secs
Recording test results
Setting GRADLE_2_4_RC_2_HOME=/home/jenkins/jenkins-slave/tools/hudson.plugins.gradle.GradleInstallation/Gradle_2.4-rc-2

**Comment by enothereska** [comment]
@dguy @mjsax have a look if you can please. Thanks.

**Comment by mjsax** [comment]
Just a meta commend: I would revert the boolean flag and only set it as last statement with try-catch (ie, set if no exception occuered). This has the advantage that it will be set in a single place and if new catch blocks get addend, we cannot forget to set it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 1]
```diff
@@ -414,8 +417,11 @@ private void shutdownTasksAndState(final boolean rethrowExceptions) {
         log.debug("{} shutdownTasksAndState: shutting down all active tasks [{}] and standby tasks [{}]", logPrefix,
             activeTasks.keySet(), standbyTasks.keySet());
 
-        // Commit first as there may be cached records that have not been flushed yet.
-        commitOffsets(rethrowExceptions);
+        // only commit under clean exit
+        if (!caughtException) {
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: I think there might be a bit more to it than this. Do we need to consider the flushing of the StateStores and the producer, too?
I.e, we have state in the state stores that is going to get flushed - that state is related to the offsets that we are not committing. That state is not-only going to be flushed to disk, but also written to the changelog topics. 
Although, without EOS i'm not sure we can do much. We can't guarantee that the state hasn't already been flushed to the store and the changelog, i.e., eviction can happen at any time.

**Review by dguy — COMMENTED** [review]
+1 on what @mjsax suggested.
Plus i left one comment that may or may not need some thought

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 1 above)
Comment (enothereska) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: @dguy as you mention, we cannot do atomic writes right now. Not committing is what will guarantee at least once, but we cannot guarantee exactly-once yet. E.g., if state store's cache is 0, the data has already gone to persistent storage. This is also what you're saying.

**Review by mjsax — APPROVED** [review]

**Comment by enothereska** [comment]
cc @guozhangwang

**Comment by guozhangwang** [comment]
Merged to trunk, thanks @enothereska !

**Pull request closed without merging**